lib/Dialect/SV/Transforms/HWExportModuleHierarchy.cpp

```diff
@@ -0,0 +1,135 @@
+  // opaque.
+  if (moduleOp) {
+    bool firstIteration = true;
+    for (auto op : moduleOp.getOps<hw::InstanceOp>()) {
+      if (firstIteration)
+        firstIteration = false;
+      else
+        os << ",";
+      printHierarchy(op, symbolTable, os);
+    }
+  }
+
+  os << "]}";
+}
+
+/// Return the JSON-serialized module hierarchy for the given module as the top.
+static std::string extractHierarchyFromTop(hw::HWModuleOp op,
+                                           SymbolTable &symbolTable) {
+  std::string resultBuffer;
+  llvm::raw_string_ostream os(resultBuffer);
+
+  // As a special case for top-level module, set instance name to module name,
+  // since the top-level module is not instantiated.
+  os << "{"
+     << "\"instance_name\":\"" << op.getName() << "\","
+     << "\"module_name\":\"" << op.getName() << "\","
+     << "\"instances\":[";
+
+  bool firstIteration = true;
+  for (auto op : op.getOps<hw::InstanceOp>()) {
+    if (firstIteration)
+      firstIteration = false;
+    else
+      os << ",";
+    printHierarchy(op, symbolTable, os);
+  }
+  os << "]}";
+
+  return resultBuffer;
+}
+
+/// Find the modules corresponding to the firrtl mainModule and DesignUnderTest,
+/// and if they exist, emit a verbatim op with the module hierarchy for each.
+void HWExportModuleHierarchyPass::runOnOperation() {
+  mlir::ModuleOp mlirModule = getOperation();
+  auto builder = OpBuilder::atBlockEnd(mlirModule.getBody());
+  SymbolTable symbolTable(mlirModule);
+
+  if (auto mainModuleName =
+          mlirModule->getAttrOfType<mlir::StringAttr>("firrtl.mainModule")) {
+    auto mainModule = symbolTable.lookup<hw::HWModuleOp>(mainModuleName);
+    assert(mainModule &&
+           "firrtl.mainModule should match the name of some HWModuleOp");
+
+    auto verbatimOp = builder.create<sv::VerbatimOp>(
+        builder.getUnknownLoc(),
+        extractHierarchyFromTop(mainModule, symbolTable));
+    verbatimOp->setAttr(
+        "output_file", hw::OutputFileAttr::get(
+                           builder.getStringAttr(""),
+                           builder.getStringAttr("testharness_hier.json"),
+                           /*exclude_from_filelist=*/builder.getBoolAttr(true),
+                           /*exclude_replicated_ops=*/builder.getBoolAttr(true),
+                           &getContext()));
+  }
+
+  for (auto op : mlirModule.getOps<hw::HWModuleOp>()) {
+    if (op->hasAttr("firrtl.DesignUnderTest")) {
+      auto verbatimOp = builder.create<sv::VerbatimOp>(
+          builder.getUnknownLoc(), extractHierarchyFromTop(op, symbolTable));
+      verbatimOp->setAttr(
+          "output_file",
+          hw::OutputFileAttr::get(
+              builder.getStringAttr(""),
+              builder.getStringAttr("module_hier.json"),
+              /*exclude_from_filelist=*/builder.getBoolAttr(true),
+              /*exclude_replicated_ops=*/builder.getBoolAttr(true),
+              &getContext()));
```

Comment:
Yes, good suggestion, although I think this would be obviated by @darthscsi's suggestion of creating a new attribute that means "dump from this module and downwards to file x", which can be lowered to from the two firrtl attributes here.